dBFT: initial multisignature adoptation

AnnaShaleva opened the pull request:
A part of #13. In this PR:
1. Add four nodes + RPC node privnet setup.
2. Add stand by committee and validators count configuration.
3. Support multisignature at the dBFT side.
4. Create blocks signature from dBFT's context (commits).
5. Move validators addresses from sealed Header's Extra part to unsealed Extra part.

At this stage single privnet is running and accepting blocks. Four nodes privnet is able to run and nodes can see each other, but no block acceptance yet:
```
anna@kiwi:~/Documents/GitProjects/bane-labs/go-ethereum$ cat privnet/node1/geth_node.log 
INFO [12-04|16:42:31.969] Enabling metrics collection 
INFO [12-04|16:42:31.970] Maximum peer count                       ETH=50 LES=0 total=50
INFO [12-04|16:42:31.971] Smartcard socket not found, disabling    err="stat /run/pcscd/pcscd.comm: no such file or directory"
INFO [12-04|16:42:31.973] Using pebble as the backing database 
INFO [12-04|16:42:31.973] Allocated cache and file handles         database=/home/anna/Documents/GitProjects/bane-labs/go-ethereum/privnet/node1/geth/chaindata cache=512.00MiB handles=2048
INFO [12-04|16:42:31.974] Opened ancient database                  database=/home/anna/Documents/GitProjects/bane-labs/go-ethereum/privnet/node1/geth/chaindata/ancient/chain readonly=true
INFO [12-04|16:42:31.974] State scheme set to already existing     scheme=hash
INFO [12-04|16:42:31.974] Set global gas cap                       cap=50,000,000
INFO [12-04|16:42:31.974] Initializing the KZG library             backend=gokzg
INFO [12-04|16:42:32.061] Allocated trie memory caches             clean=154.00MiB dirty=256.00MiB
INFO [12-04|16:42:32.062] Using pebble as the backing database 
INFO [12-04|16:42:32.062] Allocated cache and file handles         database=/home/anna/Documents/GitProjects/bane-labs/go-ethereum/privnet/node1/geth/chaindata cache=512.00MiB handles=2048
INFO [12-04|16:42:32.092] Opened ancient database                  database=/home/anna/Documents/GitProjects/bane-labs/go-ethereum/privnet/node1/geth/chaindata/ancient/chain readonly=false
INFO [12-04|16:42:32.093] Initialising Ethereum protocol           network=2,312,041,641 dbversion=<nil>
INFO [12-04|16:42:32.094]  
INFO [12-04|16:42:32.094] --------------------------------------------------------------------------------------------------------------------------------------------------------- 
INFO [12-04|16:42:32.094] Chain ID:  2312041641 (unknown) 
INFO [12-04|16:42:32.094] Consensus: dBFT (proof-of-authority) 
INFO [12-04|16:42:32.094]  
INFO [12-04|16:42:32.094] Pre-Merge hard forks (block based): 
INFO [12-04|16:42:32.094]  - Homestead:                   #0        (https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/homestead.md) 
INFO [12-04|16:42:32.094]  - Tangerine Whistle (EIP 150): #0        (https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/tangerine-whistle.md) 
INFO [12-04|16:42:32.094]  - Spurious Dragon/1 (EIP 155): #0        (https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/spurious-dragon.md) 
INFO [12-04|16:42:32.094]  - Spurious Dragon/2 (EIP 158): #0        (https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/spurious-dragon.md) 
INFO [12-04|16:42:32.094]  - Byzantium:                   #0        (https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/byzantium.md) 
INFO [12-04|16:42:32.094]  - Constantinople:              #0        (https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/constantinople.md) 
INFO [12-04|16:42:32.094]  - Petersburg:                  #0        (https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/petersburg.md) 
INFO [12-04|16:42:32.094]  - Istanbul:                    #0        (https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/istanbul.md) 
INFO [12-04|16:42:32.094]  - Muir Glacier:                #0        (https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/muir-glacier.md) 
INFO [12-04|16:42:32.094]  - Berlin:                      #0        (https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/berlin.md) 
INFO [12-04|16:42:32.094]  - London:                      #0        (https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/london.md) 
INFO [12-04|16:42:32.094]  - Arrow Glacier:               #0        (https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/arrow-glacier.md) 
INFO [12-04|16:42:32.094]  - Gray Glacier:                #0        (https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/gray-glacier.md) 
INFO [12-04|16:42:32.094]  
INFO [12-04|16:42:32.094] The Merge is not yet available for this network! 
INFO [12-04|16:42:32.094]  - Hard-fork specification: https://github.com/ethereum/execution-specs/blob/master/network-upgrades/mainnet-upgrades/paris.md 
INFO [12-04|16:42:32.094]  
INFO [12-04|16:42:32.094] Post-Merge hard forks (timestamp based): 
INFO [12-04|16:42:32.094]  
INFO [12-04|16:42:32.094] --------------------------------------------------------------------------------------------------------------------------------------------------------- 
INFO [12-04|16:42:32.094]  
INFO [12-04|16:42:32.094] Loaded most recent local block           number=0 hash=c3acdf..ebbb33 td=1 age=54y8mo2w
WARN [12-04|16:42:32.095] Failed to load snapshot                  err="missing or corrupted snapshot"
INFO [12-04|16:42:32.107] Rebuilding state snapshot 
INFO [12-04|16:42:32.108] Resuming state snapshot generation       root=5b141c..3a311a accounts=0 slots=0 storage=0.00B dangling=0 elapsed=1.544ms
INFO [12-04|16:42:32.108] Initialized transaction indexer          limit=2,350,000
INFO [12-04|16:42:32.109] Regenerated local transaction journal    transactions=0 accounts=0
INFO [12-04|16:42:32.112] Generated state snapshot                 accounts=5 slots=0 storage=235.00B dangling=0 elapsed=5.457ms
INFO [12-04|16:42:32.120] Gasprice oracle is ignoring threshold set threshold=2
WARN [12-04|16:42:32.122] Engine API enabled                       protocol=eth
WARN [12-04|16:42:32.122] Engine API started but chain not configured for merge yet 
INFO [12-04|16:42:32.122] Starting peer-to-peer node               instance=Geth/v1.13.1-stable-de8f4729/linux-amd64/go1.21.1
INFO [12-04|16:42:32.123] Stored checkpoint snapshot to disk       number=0 hash=c3acdf..ebbb33
INFO [12-04|16:42:32.169] New local node record                    seq=1,701,697,352,169 id=4fd4b8bf4c11fc93 ip=127.0.0.1 udp=30306 tcp=30306
INFO [12-04|16:42:32.170] Started P2P networking                   self=enode://a8974c6938edbedb8a830db501c7f3b631d9e91ea5b3030173ca6dd9e019fffc39107565b5af287ce1e2d8315576f79185134ea32d26e7d17bdd126f6559e6c9@127.0.0.1:30306
INFO [12-04|16:42:32.171] IPC endpoint opened                      url=/home/anna/Documents/GitProjects/bane-labs/go-ethereum/privnet/node1/geth.ipc
INFO [12-04|16:42:32.171] Generated JWT secret                     path=/home/anna/Documents/GitProjects/bane-labs/go-ethereum/privnet/node1/geth/jwtsecret
INFO [12-04|16:42:32.171] WebSocket enabled                        url=ws://127.0.0.1:8552
INFO [12-04|16:42:32.171] HTTP server started                      endpoint=127.0.0.1:8552 auth=true prefix= cors=localhost vhosts=localhost
INFO [12-04|16:42:33.037] Unlocked account                         address=0x9e14BC9c234737d9C04afEEc19948B7aA334fD4B
INFO [12-04|16:42:33.037] Legacy pool tip threshold updated        tip=0
INFO [12-04|16:42:33.037] Legacy pool tip threshold updated        tip=1,000,000,000
INFO [12-04|16:42:33.037] initializing BFT consensus 
2023-12-04T16:42:33.037+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:107	initializing dbft	{"height": 1, "view": 0, "index": 1, "role": "Primary"}
2023-12-04T16:42:33.037+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:617	reset timer	{"h": 1, "v": 0, "delay": "0s"}
INFO [12-04|16:42:33.037] Commit new sealing work                  number=1 sealhash=0912f8..144fa7 txs=0 gas=0 fees=0 elapsed="61.889µs"
2023-12-04T16:42:33.037+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:107	initializing dbft	{"height": 1, "view": 0, "index": 1, "role": "Primary"}
2023-12-04T16:42:33.037+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:617	reset timer	{"h": 1, "v": 0, "delay": "0s"}
2023-12-04T16:42:33.037+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/send.go:9	broadcasting message	{"type": "PrepareRequest", "height": 1, "view": 0}
2023-12-04T16:42:33.037+0300	INFO	dbft@v0.0.0-20230515113611-25db6ba61d5c/send.go:40	sending PrepareRequest	{"height": 1, "view": 0}
2023-12-04T16:42:33.037+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/dbft.go:617	reset timer	{"h": 1, "v": 0, "delay": "5s"}
2023-12-04T16:42:33.037+0300	DEBUG	dbft@v0.0.0-20230515113611-25db6ba61d5c/check.go:29	check preparations	{"hasReq": true, "count": 1, "M": 3}
INFO [12-04|16:42:42.194] Looking for peers                        peercount=4 tried=0 static=0

```

AnnaShaleva committed:
Makefile: add ability to setup 4-nodes+RPC privnet

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
Makefile: fix bootnode setup

The script is broken.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
config: add a notion of stand by validators

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva commented on params/config.go:
```diff
@@ -358,6 +358,8 @@ func (c *CliqueConfig) String() string {
 // DBFTConfig is the consensus engine configs for proof-of-stake based sealing.
 type DBFTConfig struct {
 	SecondsPerBlock uint64 `json:"period"` // Number of seconds between blocks to enforce
+	// TODO: extend to StandbyCommittee + ValidatorsCount.
```
@roman-khimov, I'll extend this, but later, and right now the next step is to implement the following logic: grab these validators from config and check if the consensus node in that list. If so, then sign the block, if not then don't participate in consensus. Validate all incoming blocks according to this list.

AnnaShaleva replied:
Extended.

roman-khimov commented on Makefile:
```diff
@@ -137,6 +149,31 @@ privnet_init: privnet_clean
 	@rm $(MAIN_DIR)/$(GENESIS_WORK_JSON)
 	@echo "OK! For starting use 'make privnet_start'"
 
+privnet_init_multy: privnet_clean
```
multi

AnnaShaleva replied:
Four, actually. Hope we'll add `seven` later.

roman-khimov commented on privnet/node1/node_address.txt:
```diff
@@ -1 +1 @@
-625eafa3473492007c0dd331e23b1035f6a7fb64
+f2d5f1c49048e200d16b3758cea2ff6484634b37
```
Can we avoid changing single-node keys/addresses?

roman-khimov replied:
Maybe not critical though, we don't yet have a lot of test networks.

AnnaShaleva replied:
We can avoid this by putting multi-privnet data to a separate folder, will do. Or we can even remove generator-related code, hardly ever we need to regenerate test privnet data.

AnnaShaleva replied:
OK, I made my best and I don't want to spend more time on privnet part of makefile, it sucks. This PR is the last time when we change these files (I hope). Let it be just as it is, and further (if we have time) I'm going to remove autogenerator so that we have hardcoded accounts, passwords, etc. exactly like we have in NeoGo's privnet.

AnnaShaleva committed:
Makefile: split single/four privnet setup and run

Allow to run four-nodes privnet and single-node privnet independently
and do not drop the DB between runs.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
*: add a notion of dBFT multisignature to Block's header

1. Store proper NextConsensus field in the Header's MixDigest field
   and calculated based on the next block validators (got either from
   config or from governance contract, the latter is TBD).
2. Change Header's Extra structure.
   Header's Extra structure before this commit:
   |--- extraVanity (32 byte) ---|--- signers addresses (once per epoch, x * 20 byte) or empty ---|--- single signer signature (64 byte + 1 byte) ---|
   Header's Extra structure after this commit:
   |--- extraVanity (32 byte) ---|--- validators addresses (every block, n * 20 byte) ---|--- m signer signatures (m * (64 byte + 1 byte)) ---|
3. Properly verify new Extra data against NextConsensus field.

Also this commit contains a basis for dBFT interactions with Eth API (contract calls).
This code will be moved to a separate commit.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva commented:
@roman-khimov, it's just a snippet that I'm currently working on,just to get you an idea of the code that I'm touching right now. There are several TODOs, I'll fix them tomorrow. The currently uploaded version has blocks signatures that are suitable for dBFT (at least at my POW), thus I'm asking you to pay special attention to the last commit where the Header's `Extra` structure is being changed, and tell whether this change matches our expectations or not.

The current version is able to run in single mode, the blocks are being accepted, so this is minimum working prototype with multisignatures. I need to test it with multiple nodes scenario, but it's not that easy without P2P level.

roman-khimov commented on crypto/multisig.go:
```diff
@@ -20,25 +20,39 @@ import (
+
+// GetMajorityHonestNodeCount returns minimum number of honest nodes
+// required for majority-style agreement.
+func GetMajorityHonestNodeCount(n int) int {
```
It's be nice to move it out of `crypto`, but I'm not sure what's the best package.

AnnaShaleva replied:
How about reusing neo-go util? We already have this neo-go dependency and we won't get rid of it, so why not to reuse it more.

roman-khimov replied:
It's in `smartcontract` in NeoGo and that's too much. I also hate `util` enough to avoid moving it there.

AnnaShaleva replied:
How about `dbftutil`? We already have `GetNextConsensusHash` and `GetNextConsensusScript` there.

AnnaShaleva replied:
Oh, can't be, it uses `crypto`.

AnnaShaleva replied:
Ref. https://github.com/bane-labs/go-ethereum/issues/47.

roman-khimov commented on crypto/multisig.go:
```diff
@@ -49,34 +63,35 @@ func VerifyMultiMajority(hash []byte, pubs [][]byte, sigs [][]byte) error {
 	}
 	sort.Slice(sPubs, func(i, j int) bool {
-		return bytes.Compare(sPubs[i], sPubs[j]) < 0
+		return bytes.Compare(sPubs[i][:], sPubs[j][:]) < 0
```
You can just use `==` here then.

AnnaShaleva replied:
Are you sure? It's comparator function, we need to sort them. But several lines below we can use `==`, fixed.

roman-khimov replied:
Yeah, I mean `==` lines, of course.

roman-khimov commented on core/types/block.go:
```diff
@@ -165,6 +165,18 @@ func (h *Header) EmptyReceipts() bool {
+// NextConsensus returns the next consensus address.
+func (h *Header) NextConsensus() common.Address {
+	var res common.Address
+	res.SetBytes(h.MixDigest[common.HashLength-common.AddressLength:])
```
Is it critical for us to have 20-byte value here? We could be storing the full hash otherwise because in fact it's not an address in E-chain.

AnnaShaleva replied:
Not critical at all, the only reason for this was our initial plan to use Uint160. Fixed.

roman-khimov commented on consensus/dbft/dbft.go:
```diff
@@ -635,30 +649,32 @@ func (c *DBFT) VerifyUncles(chain consensus.ChainReader, block *types.Block) err
-			if limit := uint64(len(snap.Signers)/2 + 1); seen > number-limit {
-				return errRecentlySigned
-			}
+	// TODO: fill in NextConsensus of genesis block properly or don't check it at all.
```
Must be filled with standby validators, yes.

AnnaShaleva replied:
Fixed.

roman-khimov commented on core/systemcontracts/contracts.go:
```diff
@@ -0,0 +1,8 @@
+const (
+	// GovernanceContract is a governing contract hash that manages validators
+	// voting and rewards.
+	GovernanceContract = "0x0000000000000000000000000000000000001000"
```
Aren't they configured somehow (predeployed ones)?

AnnaShaleva replied:
They are predeployed and they may be configured as far (not hard-coded). It's just the draft implementation that I've mostly ported from bnb chain, check out the [reword distribution example](https://github.com/bnb-chain/bsc/blob/1dca486622ae46636b888fa10f8a8dbf0879e7e8/consensus/parlia/parlia.go#L1003) (we may also want to port it) and [current validators retrieval](https://github.com/bnb-chain/bsc/blob/1dca486622ae46636b888fa10f8a8dbf0879e7e8/consensus/parlia/parlia.go#L1501).

roman-khimov commented on consensus/dbft/dbft.go:
```diff
@@ -1069,3 +1087,63 @@ func (c *DBFT) APIs(chain consensus.ChainHeaderReader) []rpc.API {
+}
+
+func (c *DBFT) shouldUpdateCommitteeAt(blockNum uint64) bool {
+	return blockNum%uint64(len(c.config.StandByCommittee)) == 0
```
I think we only have a notion of validators here, committee is N3-only, but this better be left as is for now and then adapted to the real governance scheme.

AnnaShaleva replied:
We can always reduce this config, it's not hard. So keep the current version for now?

roman-khimov commented on crypto/multisig.go:
```diff
@@ -49,34 +63,35 @@ func VerifyMultiMajority(hash []byte, pubs [][]byte, sigs [][]byte) error {
 		}
-		sPubs = append(sPubs, pubkey)
+
+		sPubs = append(sPubs, common.Address(Keccak256(pubkey[1:])[12:]))
```
It's a bit unobvious that this is the standard `PubkeyToAddress`, maybe worth moving it into internal function there.

AnnaShaleva replied:
Fixed. This function actually already exist, but for the "honest" public key, not its bytes.

roman-khimov commented on consensus/dbft/dbft.go:
```diff
@@ -146,29 +151,38 @@ var (
+
+// ecrecover extracts the set of validators addresses (ValidatorsCount number of them) and
+// a set of validators signatures (BFT number of them) from a signed header.
+func ecrecover(cfg *params.DBFTConfig, header *types.Header) ([]common.Address, [][]byte, error) {
```
Then it's no longer `ecrecover`, rather some `getSignersAndSigs`.

AnnaShaleva replied:
Fixed.

roman-khimov commented on consensus/dbft/dbft.go:
```diff
@@ -65,7 +70,7 @@ var (
 
 	extraVanity = 32                     // Fixed number of extra-data prefix bytes reserved for signer vanity
-	extraSeal   = crypto.SignatureLength // Fixed number of extra-data suffix bytes reserved for signer seal
+	extraSeal   = crypto.SignatureLength // Fixed number of extra-data suffix bytes reserved for a single signer seal
```
Likely this constant is just no longer relevant, `crypto.SignatureLength` is sufficient.

AnnaShaleva replied:
It's not relevant, but at the same time it's very convenient to track all the places where header's signatures are touched, thus I'd still suggest to keep it in case if in future we decide to change something in the signature-related logic. Can we?

AnnaShaleva committed:
privnet: autofill genesis Extra and MixDigest from dBFT config

Fetch validators addresses from dBFT configuration and automatically construct
genesis block's Extra and MixDigest fields. Add a check on genesis block commit
to ensure that in case of dBFT MixDigest and Extra match the expected content.
Remove any notion of custom Extra from genesis config in privnet, although manual
configuration is still possible.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva committed:
dbft: add draft implementation for interaction with contracts

That's the way how dBFT can interact with pre-deployed system contracts.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

roman-khimov commented on les/client.go:
```diff
@@ -187,6 +189,21 @@ func New(stack *node.Node, config *ethconfig.Config) (*LightEthereum, error) {
 	}
 	leth.ApiBackend.gpo = gasprice.NewOracle(leth.ApiBackend, gpoParams)
 
+	var bft *dbft.DBFT
```
Why do we need it in `les`? It should have enough data in headers to verify and it can't participate in consensus anyway.

AnnaShaleva replied:
I probably need to review block's verification scheme one more time, we shouldn't depend on it here.

AnnaShaleva replied:
OK, we actually don't use next block validators computations during block verification, so it's OK not to have contract interactions for light node.

AnnaShaleva replied:
Removed.

AnnaShaleva committed:
dbft: ensure block signature is constructed from commit payloads

This commit makes possible to honestly construct block signature from
dBFT's commit payloads.

This commit makes important change in the seal block hash: sealed
block hash includes only first 32 bytes of Extra Header's field and leaves
the rest part of Extra unhashable (validators addresses sorted by
their weights and validator's signatures of sealed part). It differs from the
previous implementation in that validators addresses are not a part of sealed
part anymore. We can safely allow this change because we have NextConsensus
field approved by dBFT process and each subsiquent block is being validated
to match parent's NextConsensus.

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva commented:
Rebased onto current bane-main.

roman-khimov commented on consensus/dbft/dbftutil/util.go:
```diff
@@ -5,11 +5,19 @@ import (
+	return common.BytesToHash(crypto.Keccak256(GetNextConsensusScript(nextBlockVals)))
+}
+
+// GetNextConsensusScript returns script containing addresses of validators. vals
```
Not a script, really.

AnnaShaleva replied:
Fixed.

roman-khimov commented on consensus/dbft/dbftutil/util.go:
```diff
@@ -5,11 +5,19 @@ import (
+
+// GetNextConsensusScript returns script containing addresses of validators. vals
+// must be sorted by their consensus weight.
+func GetNextConsensusScript(vals []common.Address) []byte {
```
FlattenAddresses?

AnnaShaleva replied:
Fixed.

roman-khimov commented on core/genesis.go:
```diff
@@ -37,6 +38,7 @@ import (
 	"github.com/ethereum/go-ethereum/params"
 	"github.com/ethereum/go-ethereum/rlp"
 	"github.com/ethereum/go-ethereum/trie"
+	"github.com/nspcc-dev/neo-go/pkg/smartcontract"
```
Please, don't import it. It's very very Neo-specific and we need just two lines of code.

AnnaShaleva replied:
Ref. https://github.com/bane-labs/go-ethereum/issues/47.

AnnaShaleva commented:
Not ready, I need to fix the tests.

AnnaShaleva committed:
protocols: fix dbft pool tests

Signed-off-by: Anna Shaleva <shaleva.ann@nspcc.ru>

AnnaShaleva commented:
Ready for review.

roman-khimov commented on consensus/dbft/change_view.go:
```diff
@@ -0,0 +1,47 @@
+// EncodeBinary implements the io.Serializable interface.
+func (c *changeView) EncodeBinary(w *io.BinWriter) {
+	w.WriteU64LE(c.timestamp)
+	w.WriteB(byte(c.reason))
```
Huh, no RLP fun?

AnnaShaleva replied:
It's just a draft, it's likely to be changed after integration with P2P.

roman-khimov commented on consensus/dbft/prepare_request.go:
```diff
@@ -0,0 +1,92 @@
+		w.WriteBytes(p.transactionHashes[i][:])
+	}
+	//	p.sealingProposal
+
```
Wow.

AnnaShaleva replied:
Not for this PR.

AnnaShaleva commented:
I've removed payloads-related data, it's an accidental change.
